rp2040: avoid device-mode state machine hang

P33M opened the pull request:
Please see the commit text for a description of the change.

Note that I now export hw_endpoint_start_next_buffer() and hw_endpoint_lock_update() in the header file, and because both the IRQ and driver code are doing read-modify-write operations the endpoint lock really should be implemented.

hathach commented:
can you provide context and reason for this PR ?

P33M commented:
This avoids a bug in rp2040. The device controller stops responding to host packets if it hits the condition described in the commit.

The state is effectively unrecoverable without a complete peripheral reset (not a USB bus reset) so it is far better to avoid it in the first place - host drivers do not respond well to a device that suddenly disappears from the bus.

hathach commented:
thanks for explaining, do we have an document for this bug in rp2040 i.e is it an errata or a note somewhere in the rp2040 document. If possible, could you detail how is your hw setup and step to reproduce the issue if possible. I would like to try out the before and after pr to see if that would fix the issue and/or making change update later on without breaking this fix.

P33M commented:
An errata note will be published in due course.

I have pushed a testcase to https://github.com/P33M/rp2040-e15-testcase - this is an SDK-compatible widget that defines a vendor-specific interface using TinyUSB. Bulk OUT packets are discarded, bulk IN packets are pseudorandomly sized and submitted to the hardware at arbitrary times.
The expected behaviour with the included python script is that the transfer will continue forever, but with a pico plugged into a Pi 4 directly, will crash after a few seconds.

P33M commented:
I don't know why the CI build is failing. It builds fine locally under CMake.

P33M commented:
Please merge - we consider it important enough that we are doing a point release of the Pico C SDK to pick up this change.

hathach commented:
> Please merge - we consider it important enough that we are doing a point release of the Pico C SDK to pick up this change.

In this case, I will try to review the code changes without hw testing as soon as I could in the next couple of days. And will try to do hand-on testing later on.

kilograham commented on src/portable/raspberrypi/rp2040/rp2040_usb.c:
```diff
@@ -57,6 +55,36 @@ TU_ATTR_ALWAYS_INLINE static inline bool is_host_mode(void)
+   * when it is expecting an ACK.
+   */
+  now = time_us_32();
+  if (now < (last_sof + 800) || now > (last_sof + 998)) {
```
i would do
```
delta =  time_us_32() - last_sof
```
to avoid rollover issues (not a biggy but still)

P33M replied:
Rollover would mean the time check would not apply for one frame in every 4.29 million, which is arguably a bug.

P33M commented on src/portable/raspberrypi/rp2040/dcd_rp2040.c:
```diff
@@ -247,16 +249,35 @@ static void __tusb_irq_path_func(reset_non_control_endpoints)(void)
+      last_sof = time_us_32();
+
+      // Kick any deferred nonperiodic IN endpoints
+      for (i = 0; i < USB_MAX_ENDPOINTS; i++) {
```
True, most applications will use an average of 2 low-numbered EPs for bulk.

hathach commented on src/portable/raspberrypi/rp2040/dcd_rp2040.c:
```diff
@@ -390,6 +412,9 @@ void dcd_init (uint8_t rhport)
   usb_hw->inte     = USB_INTS_BUFF_STATUS_BITS | USB_INTS_BUS_RESET_BITS | USB_INTS_SETUP_REQ_BITS |
                      USB_INTS_DEV_SUSPEND_BITS | USB_INTS_DEV_RESUME_FROM_HOST_BITS |
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+                     USB_INTS_DEV_SOF_BITS |
```
SOF should only be enabled when needed.

hathach commented on src/portable/raspberrypi/rp2040/dcd_rp2040.c:
```diff
@@ -247,16 +249,36 @@ static void __tusb_irq_path_func(reset_non_control_endpoints)(void)
+          hw_endpoint_lock_update(ep, -1);
+        }
+      }
+      pending_nonperiodic_xfer = 0;
```
we should only clear what we handled i.e `parked_eps`

```C
pending_nonperiodic_xfer &= ~parked_eps
```

However, since pending_nonperiodic_xfer is accessed by both dcd_edpt_xfer() and sof isr. There is still chance for race condition since both try to read-then-set the variable.

P33M replied:
Updated to guard the entire section with the endpoint lock. Note that the loop exits if parked_eps = 0, and pending_nonperiodic_xfer is volatile. So taking a local copy avoids emitting tons of loads/stores inside the for loop.

P33M replied:
Reimplemented in v3

hathach commented on src/portable/raspberrypi/rp2040/rp2040_usb.c:
```diff
@@ -218,8 +246,14 @@ void hw_endpoint_xfer_start(struct hw_endpoint *ep, uint8_t *buffer, uint16_t to
+    hw_endpoint_start_next_buffer(ep);
+  } else {
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+    pending_nonperiodic_xfer |= (uint16_t)(1 << tu_edpt_number(ep->ep_addr));
```
`pending_nonperiodic_xfer` are both read-then-set in both here via dcd_edpt_xfer() and sof isr. There is a race condition e.g 
- we need to set with 0x02 mask, after the read here when pending_nonperiodic_xfer = 0x01
- sof isr occurred, process the 0x01 and clear the bit, sof return
- pending_nonperiodic_xfer is set with 0x03 therefore incorrectly schedule an redundant transfer.

P33M replied:
Accesses to this variable are now guarded by the endpoint lock (which implements a pico-sdk critical_section). Would you prefer folding the variable into the hw_endpoint array?

hathach replied:
thank you for adding the guard, I think we should move the variable to hw_endpoint array, since the tinyusb stack only submit 1 transfer per endpoint at a time. This will remove the need to run critical section which is an rather expensive function.

hathach commented on src/portable/raspberrypi/rp2040/rp2040_usb.c:
```diff
@@ -329,15 +363,15 @@ bool __tusb_irq_path_func(hw_endpoint_xfer_continue)(struct hw_endpoint *ep)
   else
   {
-    _hw_endpoint_start_next_buffer(ep);
+    hw_endpoint_start_next_buffer(ep);
```
rp2040_critical_frame_period() should also be checked here as well, since this will also trigger an usb transaction. I think this place will likely to cause the issue than one in dcd_edpt_xfer() since it schedule packet back-to-back here.

hathach reviewed (commented):
late review since I was busy with other works. After going through the change, although the pr fix a severe issue, however due to the impact of this PR in performance i.e 20% off bandwidth and force to have sof (which will hurt power consumption). I would like to do actual hand-on testing with the PR before merging. Though I couldn't do that until another week. 

Meanwhile there is a few issues I think we should fix first
1. Since this only occurs with bulk in transfer, we should only enable SOF when needed i.e when there is active bulk-in within 20% last frame. This will minimize the impact on current consumption. Since not all device make use of Bulk in endpoint e.g HID device. And even with MSC/cdc device, it is more often than not, device sitting idle than transferring data. and when it does, it is more often within first 80% frame.
2. hw_endpoint_xfer_continue() should also check of critical frame i.e all places where hw_endpoint_start_next_buffer() is called to trigger usb transaction
3. There is a race condition with `pending_nonperiodic_xfer` since both dcd_edpt_xfer() and sof isr try to read-then-set it. We should use an bit in hw_endpoint to address this so that each endpoint is set and clear invidually.

Let me know if above request makes sense and feel free to provide feedback on them. I am open to all discussion.

hathach commented:
Update: SOF is only used to sync time, maybe we don’t need to sync it every frame, and could only need to sync every 1000ms or so. Since sof is very consistent from host.

P33M commented:
> Meanwhile there is a few issues I think we should fix first
> 
> 1. Since this only occurs with bulk in transfer, we should only enable SOF when needed i.e when there is active bulk-in within 20% last frame. This will minimize the impact on current consumption. Since not all device make use of Bulk in endpoint e.g HID device. And even with MSC/cdc device, it is more often than not, device sitting idle than transferring data. and when it does, it is more often within first 80% frame.

I can address this by making SOF enabled iff pending_nonperiodic_xfer != 0, i.e. we need to pay attention to SOF. The buffer enqueue check will filter out endpoint types that don't need the fix.

> 2. hw_endpoint_xfer_continue() should also check of critical frame i.e all places where hw_endpoint_start_next_buffer() is called to trigger usb transaction
> 3. There is a race condition with `pending_nonperiodic_xfer` since both dcd_edpt_xfer() and sof isr try to read-then-set it. We should use an bool/flag in hw_endpoint to address this so that each endpoint is set and clear individually.

These two points mean that a proper critical section implementation needs to be added to hw_endpoint_lock_update(). Which of the pico-sdk synchronisation primitives would be best to use in tinyUSB's case?

> PS: I have one extra question does the 20% frame take in account of the double buffered i.e
> 
> * if we are about to schedule a double buffered transfer (buf0 and buf1) which result in 2 consecutive packet, should the critical frame now is increase to 40% ?
> * or if we are about to schedule a single buffer transfer, should the critical frame is reduced to 10% ?

I believe the testcase exercises the double-buffered behaviour.  The 20% reservation should also be enough to cover the cases where you get:
- Two or more IN endpoints activated at the same time before the reservation period, whether or not they are double-buffered
- An OUT scheduled by the HC before an IN transfer, and the IN was delayed by the hub's TT due to the large OUT packet.

kilograham reviewed (commented):
one thing i had been vaguely wondering about before this change is why the existing methods have +1, -1; this implies to me that re-entrancy is required, however this implementation does not support it.

If re-entrancy is not required, I think we should make that explicit.

hathach reviewed (commented):
sorry for late response again, the Lunar New Year aka TET in my country forcing me to clean my house/desk and leave not much time for the work. I don't think I will have time to test this out on actual RPI4 in another couple of weeks, and this has been on pending for too long (my bad). If you could update the PR 
- to move the parked_ep into hw_endpoint array without the use of critical_section
- comment out/revert the hw_endpoint_lock_update() implementation since there is other places that use spinlock as well, Some is not neccessary since usbd only schedule 1 transfer per endpoint until it is complete. We could re-add spinlock later on with more closer look.

Afterwards, I think we could merge this as it is. I will do the tests after merged and make changes when & where needed with PR mentioned you for review.

P33M commented:
I've addressed the review comments and updated the PR but unfortunately it looks like the SOF interrupt has to be constantly enabled. I cache the time of the previous SOF event and compute delta modulo 1000, but this will fail after several seconds as it appears there is enough clock drift to push the software blanking period away from the critical period.

We have no other way of keeping a stable count of where we are in a frame without logging every interrupt so whenever a bulk IN endpoint is active SOF must also be active.

kilograham commented on src/portable/raspberrypi/rp2040/rp2040_usb.h:
```diff
@@ -11,11 +11,20 @@
+#endif
+
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+#define PICO_RP2040_USB_FAST_IRQ 1
```
if we want to unconditionally set it to 1, we should probably #undef it first

P33M replied:
Yes, compilation will fail if someone's set this already

kilograham replied:
which i think they may have

kilograham commented on src/portable/raspberrypi/rp2040/rp2040_usb.h:
```diff
@@ -74,11 +90,24 @@ typedef struct hw_endpoint
+extern volatile uint32_t last_sof;
+#endif
+
+uint32_t rp2040_critical_frame_period(struct hw_endpoint *ep);
```
maybe consider making this a macro that returns 0 when !TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX rather than having an external function that always returns false.

kilograham commented on src/portable/raspberrypi/rp2040/dcd_rp2040.c:
```diff
@@ -249,11 +246,28 @@ static void __tusb_irq_path_func(dcd_rp2040_irq)(void)
       handled |= USB_INTF_DEV_SOF_BITS;
 
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+    hw_endpoint_lock_update(ep, 1);
```
looks like indent here is wrong

P33M replied:
The whole file uses 2 and 4-space indents interchangeably throughout, but yes it looks off

kilograham commented on src/portable/raspberrypi/rp2040/dcd_rp2040.c:
```diff
@@ -447,6 +461,15 @@ void dcd_sof_enable(uint8_t rhport, bool en)
 
   _sof_enable = en;
 
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
```
should this be in the `if (en)` branch

kilograham replied:
oh, i guess not

kilograham commented on src/portable/raspberrypi/rp2040/rp2040_usb.h:
```diff
@@ -64,7 +76,11 @@ typedef struct hw_endpoint
     // Interrupt, bulk, etc
     uint8_t transfer_type;
-    
+
```
since your here, how about moving `active` from above with suitable comment into this group of 1 byte fields... we have 32 of these structures and we were wasting 4 bytes per.

P33M replied:
```
(gdb) ptype /o struct hw_endpoint
/* offset    |  size */  type = struct hw_endpoint {
/*    0      |     1 */    _Bool configured;
/*    1      |     1 */    _Bool rx;
/*    2      |     1 */    uint8_t ep_addr;
/*    3      |     1 */    uint8_t next_pid;
/*    4      |     4 */    io_rw_32 *endpoint_control;
/*    8      |     4 */    io_rw_32 *buffer_control;
/*   12      |     4 */    uint8_t *hw_data_buf;
/*   16      |     1 */    _Bool active;
/* XXX  1-byte hole  */
/*   18      |     2 */    uint16_t remaining_len;
/*   20      |     2 */    uint16_t xferred_len;
/* XXX  2-byte hole  */
/*   24      |     4 */    uint8_t *user_buf;
/*   28      |     2 */    uint16_t wMaxPacketSize;
/*   30      |     1 */    uint8_t transfer_type;
/*   31      |     1 */    uint8_t pending;

                           /* total size (bytes):   32 */
                         }
```
three according to gdb, but yes holes are bad

kilograham replied:
ah, ok, i thought it would make it smaller, but i was thinking of the 2 uint16_t s as being 4 byte aligned for some reason :-(

P33M replied:
In host mode the two extra fields end up growing the structure to 36B, so it is worth rearranging so that pad bytes at the end of the structure get used instead.

hathach reviewed (approved):
Thank you very much for the PR and your patient. I still haven't got time to tried this out on actual hardware described with Errata 15 (with Pi4 USB Hub). But it looks great now. 

I have made some minor changes mostly whitespace, rename, and minor refactor. Let me know these changes made sense to you. If all is good, I think we could merge this as it is.

lurch commented:
@hathach I think your comments mentioning "walkaround" / "walk-around" should actually say "workaround" ?

hathach commented:
> @hathach I think your comments mentioning "walkaround" / "walk-around" should actually say "workaround" ?

Indeed, that is my bad

hathach commented:
> I have no objections to the edits.

perfect !! Thank you very much, merged now. Feel free to make more follow up PR should this needs more changes to fix.
